Alchemy

AlexPunches opened the pull request:
Добавил алхимию и постгресс, но без ассинхронности
Собрал модели и pydantic

Чтобы потестировать нужно запустить докер контейнер
`docker-compose -f docker-compose.dev.yaml up`

и запустить сервер 
`uvicorn sql_app.main:app`

можно пощелкать по openapi [http://127.0.0.1:8000/docs](http://127.0.0.1:8000/docs)

уже есть пара ендпоинтов и несколько базовых функций

AlexPunches committed:
Add postgress and sqlalchemy moodels

tables Users, Cohorts, Faculty

AlexPunches committed:
Add SQLAlchemy and PostgreSQL

Модуль в sql_app
Добавлены модели, pydantic, базовые функции
и несколько эндпоинтов потестировать

на http://127.0.0.1:8000/docs можно немного потестировать

AlexPunches committed:
merge develop from repo

AntonZelinsky commented on docker-compose.dev.yaml:
```diff
@@ -0,0 +1,13 @@
+    ports:
+      - 35432:5432
+    volumes:
+      - ./data:/var/lib/postgresql/data
```
Я думаю, нет смысла делать папку локально. 
Используй такую конструкцию для создания тома: `- postgres_data:/var/lib/postgresql/data/`

AntonZelinsky replied:
Тогда и в гит игноре не понадобится добавлять папку.
P.S. А что в этой папке за файлы находятся? Мне всегда было интересно, в каком виде постгрес хранит свои данные)

AlexPunches replied:
файлы постгресс-данных переместил из каталога приложения.

там в пострессе куча папок и файлов, там целый мир.
разобраться в них это наврено отдельный курс должен быть на ЯП :)
https://postgrespro.ru/docs/postgresql/13/storage-file-layout

базу поправить получилось легко, alembic не пригодились,
просто удалил старую ;)
данных всеравно пока небыло

AntonZelinsky commented on sql_app/crud.py:
```diff
@@ -0,0 +1,67 @@
+    return db_user
+
+
+def get_faculties(db: Session, skip: int = 0, limit: int = 100):
+    return db.query(models.Faculty).offset(skip).limit(limit).all()
+
+
+def get_faculty_by_name(db: Session, name: str):
+    return db.query(models.Faculty).filter(models.Faculty.name == name).first()
+
+
+def create_faculty(db: Session, faculty: schemas.FacultyCreate):
+    db_faculty = models.Faculty(name=faculty.name)
+    db.add(db_faculty)
+    db.commit()
+    db.refresh(db_faculty)
+    return db_faculty
+
+
+def get_cohort_by_faculty_id_and_number(db: Session, faculty_id: int,
+                                        number: int, ):
+    return db.query(models.Cohort).filter(
+        models.Cohort.faculty_id == faculty_id,
+        models.Cohort.number == number,
+    ).first()
+
+
+def get_cohorts(db: Session, skip: int = 0, limit: int = 100):
+    return db.query(models.Cohort).offset(skip).limit(limit).all()
+
+
+def create_cohort(db: Session, cohort: schemas.CohortCreate):
+    db_cohort = models.Cohort(
+        number=cohort.number,
+        faculty_id=cohort.faculty_id,
+    )
+    db.add(db_cohort)
+    db.commit()
+    db.refresh(db_cohort)
+    return db_cohort
```
Ты оверхедишь! Сам придумываешь требования и реализовываешь их..(
У Яндекса есть такие майки, они значат, что проджект менеджер к которому приходят разработчики не их заказчик и он не может посоветовать как лучше. Это нужно узнавать у заказчика, что он хочет и что имел в виду под какой-то неочевидной формулировкой. Потому что когда реализовывают фитчу и отдают заказчику, а оказывается что это не то что он хотел, возникает много вопросов.
![image](https://user-images.githubusercontent.com/1431279/137233826-5cd48153-a2fd-40fa-a830-184065e4eae8.png)

Раз сделал, то пусть будет, но всё же, нужно идти по шагам!

AntonZelinsky replied:
Потом вспомнил, что мы же делаем не апи к сайту, а бота, поэтому, оставлять чтобы было не будем

AlexPunches replied:
Прикольная футболка :)

Удалил лишнее, оставил только добавить юзера и когорту, ну и гет_юзер и гет_когорту соответственно

AntonZelinsky commented on sql_app/models.py:
```diff
@@ -0,0 +1,37 @@
+    faculty = relationship("Faculty", back_populates="cohorts")
+
+
+class Faculty(Base):
```
Давай всё таки удалим эту таблицу, так как от этого будет много лишних телодвижений, а расшаривать на другие факультеты мы пока не планируем.

AntonZelinsky commented on sql_app/main.py:
```diff
@@ -0,0 +1,71 @@
+
+
+@app.get("/faculties/", response_model=List[schemas.Faculty])
+def read_faculties(
```
Сделай один эндпоинт который добавляет данные в базу, другой который получает. Всё остальное можно удалить..( Мы не будем использовать такие способы работы с АПИ. Я бы рассказал что правильнее взять, но ты запилил пол приложения за вечер, я не успел.)) 
Таким образом пишут только тогда, когда там планируется 1-2 эндпоинта и ничего больше.
Да и вообще, нам не нужны эти апишки сейчас, мы же не планируем делать фронт в ближайшее время.

AntonZelinsky replied:
Валидации соответственно тоже не нужны.(( 
Я рассказывал уже про поддержание неиспользуемого кода и почему это плохо

AlexPunches replied:
сделал один для когорты 
и один для юзера

согласен, фронт на данном этапе совсем не нужен
через openapi.json можно добавить все когорты,
а юзеры уже через телегу будут добавляться.

AntonZelinsky reviewed (changes requested):
Будет тяжело, но придётся много чего удалить..(

AntonZelinsky commented:
Когда начнёшь править базу, скорее всего у тебя начнёт падать приложение с ошибками из-за того, что модель данных поменялась. Тут нужно подключать библиотеку alembic, но по хорошему, не нужно делать в этом ПР, потому что ты уже насоздвал много разных папок и файлов, и ещё с десяток таких же только запутают всё. В следующем ПР добавишь.

AntonZelinsky commented:
Вообще, структура папок будет приблизительно такой https://fastapi.tiangolo.com/tutorial/bigger-applications/
Но скорее всего, создадим ещё модуль core, где будут хранится штуки связанные с базой и которые будут шарится между двумя приложениями.

AlexPunches committed:
Fix PR-7. Remove extra code

Остались две модели: User и Cohort
данные постгресса теперь не в папке
осталось два ендпоинта /users/ и /cohorts/

AlexPunches committed:
fix schemas for add Cohort with notion_db_id

AntonZelinsky commented on sql_app/models.py:
```diff
@@ -0,0 +1,26 @@
+from .database import Base
+
+
+class User(Base):
```
Создай папку core  и положи в него этот файл. Это общие модели между фронтом и ботом

AntonZelinsky commented on sql_app/main.py:
```diff
@@ -0,0 +1,64 @@
+from typing import List
```
Папка  sql_app имеет плохое название. Создай папку app и файл _ _ init__.py и перенеси туда это содержимое

AntonZelinsky replied:
Так же, в папке создать файл апи и в него поместить все роуты и регистрировать уже не на апп декоратор, а на 
апи роутер.
подключив предварительно такую штуку `router = APIRouter(tags=["Blogs"], prefix="/blog")`
Можно почитать тут https://fastapi.tiangolo.com/tutorial/bigger-applications/#apirouter

AntonZelinsky replied:
А роутер регистрировать уже после создания приложения

AntonZelinsky commented on sql_app/crud.py:
```diff
@@ -0,0 +1,38 @@
+from sqlalchemy.orm import Session
```
Этот файл корректнее уже назвать как сервис и положить в папку апп

AntonZelinsky commented on sql_app/main.py:
```diff
@@ -0,0 +1,64 @@
+
+models.Base.metadata.create_all(bind=engine)
+
+app = FastAPI()
```
Хорошо бы создавать приложение сразу в методе, для того чтобы конфигурировать его там и при повторном импорте не запускать приложение.
Пример привёл ниже, только назвать его лучше create_app и не создавать экземпляр в этом файле, пусть создаётся в корне приложения
https://github.com/nsidnev/fastapi-realworld-example-app/blob/master/app/main.py#L13

AlexPunches replied:
сделал в методе create_app()

но вот не могу сообразить как создавать енкземляр в корне приложения, 
у нас сейчас там бот создается, и их надо как-то распораллелить

AntonZelinsky commented:
Глянь хорошую статейку, можно пару интересных архитектурный идей узнать https://testdriven.io/courses/fastapi-celery/app-factory/

AlexPunches committed:
refactoring architectures for router, etc

AntonZelinsky commented on docker-compose.dev.yaml:
```diff
@@ -0,0 +1,16 @@
+version: '3.9'
```
Создай файл readme и напиши что это за штука и как её запустить. Минимальная инструкция.

AlexPunches replied:
сделал инструкцию как запустить докер и два эндпоинта.

а когда обьединим всю архитектуру, тогда соответственно редми перепишем

AntonZelinsky commented on sql_app/database.py:
```diff
@@ -0,0 +1,21 @@
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy.orm import sessionmaker
+
+SQLALCHEMY_DATABASE_URL = \
```
давай в кор тоже перенесём эту штуку. так как и бот и апи будет им пользоваться.

AntonZelinsky commented on sql_app/schemas.py:
```diff
@@ -0,0 +1,41 @@
+from datetime import datetime
```
А схему в апп перенесём, так как она касается только апи

AntonZelinsky reviewed (commented):
Ещё есть пару мелких вопросов, но уже поправим их в других ПРах.

AlexPunches committed:
Add README.md

Пока в редми инструкция для запуска и тестирования эндпоинтов
в процессе обьединения архитектуры будем менять инструкцию

AlexPunches committed:
Modernization architecture

AlexPunches committed:
Update README.md

AntonZelinsky commented on README.md:
```diff
@@ -0,0 +1,50 @@
+#Квик старт
```
Почему-то заголовки не подтянулись. Скорее всего из-за того, что нет пробела после шарпа.

AntonZelinsky commented on README.md:
````diff
@@ -0,0 +1,50 @@
+Добавить переменные окружения
+```bash
+$ mv .env.example .env
+$ vim .env
````
Не пиши про вим. Лучше написать что нужно отредактировать этот файл и всё. 
Я например до сих пор не знак из его выйти! Мне недавно мало что не пришлось перезагружать сервер, из-за того, что я не мог выйти из него!) Я решил узнать, установлен ли вим на сервер по дефолту и это было ошибкой) Мне пришлось принудительно закрыть сессию с сервером. Гугл никак не помогал выйти из его..(

Да и пользователям винды он ничего не скажет

AlexPunches replied:
У меня тоже с вимом непростое знакомство было :)

а вот про пользователей виндовсов совсем не подумал. Виноват. Исправлюсь :)

AntonZelinsky commented on core/database.py:
```diff
@@ -0,0 +1,21 @@
+from sqlalchemy.orm import sessionmaker
+
+SQLALCHEMY_DATABASE_URL = \
+    "postgresql://root:root@localhost:35432/schedule_notifier"
```
АААА, что за хардкод???

AlexPunches replied:
Как раз на этом ПР зачитывался 12-factor App
а третий фактор один из немногих, который более менее понял :-)

AntonZelinsky commented on README.md:
````diff
@@ -0,0 +1,54 @@
+$ vim .env
+```
+
+Запустить докер-контейнер
````
Я думаю, тут стоит написать для чего нужен докер, чтобы развернуть локальную базу данныхв докере, но он вовсе не обязателен. И по хорошему, конфиг подключения к базе можно добавить в экзампл, поскольку мы его предлагаем в купе с приложением.
